File: backend/app/services/catalog.py
import math

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.models import (
    GarmentType,
    MediaAsset,
    PricingRule,
    Product,
    ProductImage,
    ProductVariant,
    Theme,
)
from app.repositories.catalog import ProductPage, ProductQuery, get_color, get_sizes, get_theme
from app.repositories.catalog import get_product_by_slug as repository_get_product_by_slug
from app.repositories.catalog import list_products as repository_list_products
from app.schemas.catalog import (
    CatalogOptionsResponse,
    GarmentOptionsResponse,
    ImageResponse,
    OptionResponse,
    PriceResponse,
    ProductCreate,
    ProductImagesResponse,
    ProductListResponse,
    ProductResponse,
    ThemeResponse,
    VariantResponse,
)
from app.services.media import media_url_resolver


class CatalogValidationError(ValueError):
    pass


def _prices_for(session: Session, garment_id: int, kind: str) -> dict:
    rules = session.scalars(
        select(PricingRule).where(
            PricingRule.garment_type_id == garment_id,
            PricingRule.product_kind == kind,
        )
    )
    return {
        rule.currency: PriceResponse(minorAmount=rule.amount_minor, currency=rule.currency)
        for rule in rules
    }


def _image_response(image: ProductImage) -> ImageResponse:
    return ImageResponse(
        src=media_url_resolver.resolve(image),
        alt=image.alt_text,
        width=image.width,
        height=image.height,
    )


def serialize_product(session: Session, product: Product) -> ProductResponse:
    primary = next((image for image in product.images if image.role == "primary"), None)
    if primary is None:
        raise RuntimeError(f"Published product {product.id} has no primary image")
    hover = next((image for image in product.images if image.role == "hover"), None)
    gallery = [image for image in product.images if image.role == "gallery"]
    prices = _prices_for(
        session,
        product.garment_type_id,
        "customized" if product.category == "customized" else "standard",
    )
    variants = []
    for variant in sorted(product.variants, key=lambda item: item.size.display_order):
        overrides = {
            item.currency: PriceResponse(minorAmount=item.amount_minor, currency=item.currency)
            for item in variant.prices
        }
        variants.append(
            VariantResponse(
                id=variant.id,
                colorId=product.color.slug,
                sizeId=variant.size.slug,
                availability=variant.availability,
                prices=overrides or None,
            )
        )
    return ProductResponse(
        id=product.id,
        slug=product.slug,
        name=product.name,
        category=product.category,
        garmentType=product.garment_type.slug,
        theme=(
            ThemeResponse(id=product.theme.slug, label=product.theme.name)
            if product.theme
            else None
        ),
        description=product.description,
        images=ProductImagesResponse(
            primary=_image_response(primary),
            hover=_image_response(hover) if hover else None,
            gallery=[_image_response(image) for image in gallery],
        ),
        colors=[
            OptionResponse(
                id=product.color.slug, label=product.color.name, swatch=product.color.swatch
            )
        ],
        sizes=[
            OptionResponse(id=variant.size.slug, label=variant.size.name)
            for variant in sorted(product.variants, key=lambda item: item.size.display_order)
        ],
        prices=prices,
        variants=variants,
        availability=product.availability,
        displayOrder=product.display_order,
    )


def list_catalog_products(session: Session, query: ProductQuery) -> ProductListResponse:
    page: ProductPage = repository_list_products(session, query)
    return ProductListResponse(
        items=[serialize_product(session, product) for product in page.items],
        page=page.page,
        pageSize=page.page_size,
        total=page.total,
        pages=math.ceil(page.total / page.page_size) if page.total else 0,
    )


def get_catalog_product(session: Session, slug: str) -> ProductResponse:
    product = repository_get_product_by_slug(session, slug)
    if product is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Product not found")
    return serialize_product(session, product)


def get_catalog_options(session: Session) -> CatalogOptionsResponse:
    garments = list(
        session.scalars(
            select(GarmentType)
            .options(selectinload(GarmentType.colors), selectinload(GarmentType.sizes))
            .order_by(GarmentType.id)
        )
    )
    themes = list(session.scalars(select(Theme).order_by(Theme.id)))
    return CatalogOptionsResponse(
        themes=[ThemeResponse(id=theme.slug, label=theme.name) for theme in themes],
        garments=[
            GarmentOptionsResponse(
                id=garment.slug,
                label=garment.name,
                colors=[
                    OptionResponse(id=color.slug, label=color.name, swatch=color.swatch)
                    for color in garment.colors
                ],
                sizes=[OptionResponse(id=size.slug, label=size.name) for size in garment.sizes],
                standardPrices=_prices_for(session, garment.id, "standard"),
                customizedPrices=_prices_for(session, garment.id, "customized"),
            )
            for garment in garments
        ],
    )


def resolve_product_configuration(
    session: Session, payload: ProductCreate, current_product_id: str | None = None
):
    garment = session.scalar(
        select(GarmentType)
        .where(GarmentType.slug == payload.garment_slug)
        .options(selectinload(GarmentType.colors), selectinload(GarmentType.sizes))
    )
    if garment is None:
        raise CatalogValidationError("Unknown garment type")
    if payload.category != "customized" and payload.category != garment.slug:
        raise CatalogValidationError("Standard product category must match its garment type")
    if payload.category == "customized" and payload.theme_slug:
        raise CatalogValidationError("Customized products cannot use predefined themes")
    if payload.category != "customized" and not payload.theme_slug:
        raise CatalogValidationError("Standard products require a theme")

    theme = get_theme(session, payload.theme_slug) if payload.theme_slug else None
    if payload.theme_slug and theme is None:
        raise CatalogValidationError("Unknown theme")
    color = get_color(session, payload.color_slug)
    if color is None or color.id not in {item.id for item in garment.colors}:
        raise CatalogValidationError("Color is not available for this garment")

    requested_sizes = set(payload.size_slugs)
    sizes = get_sizes(session, requested_sizes)
    allowed_size_ids = {item.id for item in garment.sizes}
    if len(sizes) != len(requested_sizes) or any(size.id not in allowed_size_ids for size in sizes):
        raise CatalogValidationError("Size is not available for this garment")
    slug_statement = select(Product.id).where(Product.slug == payload.slug)
    if current_product_id:
        slug_statement = slug_statement.where(Product.id != current_product_id)
    if session.scalar(slug_statement):
        raise CatalogValidationError("Product slug already exists")
    if sum(image.role == "primary" for image in payload.images) != 1:
        raise CatalogValidationError("A product requires exactly one primary image")
    if len({image.position for image in payload.images}) != len(payload.images):
        raise CatalogValidationError("Image positions must be unique")
    if any(
        not image.media_asset_id and not image.storage_key and not image.public_url
        for image in payload.images
    ):
        raise CatalogValidationError("Every image requires a storage reference")
    return garment, theme, color, sorted(sizes, key=lambda item: item.display_order)


def build_product_images(session: Session, payload: ProductCreate) -> list[ProductImage]:
    images = []
    for image in payload.images:
        asset = session.get(MediaAsset, image.media_asset_id) if image.media_asset_id else None
        if image.media_asset_id and asset is None:
            raise CatalogValidationError("Unknown media asset")
        images.append(
            ProductImage(
                media_asset=asset,
                storage_key=asset.storage_key if asset else image.storage_key,
                public_url=(
                    asset.public_url
                    if asset
                    else (str(image.public_url) if image.public_url else None)
                ),
                alt_text=image.alt_text,
                width=asset.width if asset else image.width,
                height=asset.height if asset else image.height,
                position=image.position,
                role=image.role,
            )
        )
    return images


def create_product(session: Session, payload: ProductCreate) -> Product:
    """Internal mutation foundation. It is intentionally not exposed by public routes."""
    garment, theme, color, sizes = resolve_product_configuration(session, payload)

    product = Product(
        name=payload.name,
        slug=payload.slug,
        description=payload.description,
        category=payload.category,
        garment_type=garment,
        theme=theme,
        color=color,
        availability=payload.availability,
        display_order=payload.display_order,
    )
    product.variants = [
        ProductVariant(size=size, availability=payload.availability)
        for size in sorted(sizes, key=lambda item: item.display_order)
    ]
    product.images = build_product_images(session, payload)
    session.add(product)
    session.flush()
    return product

File: backend/src/services/catalog.js
import crypto from "node:crypto";
import { GarmentType, PricingRule, Theme } from "../models/reference.js";
import { MediaAsset } from "../models/media.js";
import { Product } from "../models/product.js";
import { HttpError } from "../middleware/errors.js";
import { env } from "../config/env.js";

const priceResponse = (rules) => Object.fromEntries(rules.map((r) => [r.currency, { minorAmount: r.amountMinor, currency: r.currency }]));
const imageResponse = (image) => ({ src: image.publicUrl || `${env.PUBLIC_API_URL.replace(/\/$/, "")}/media/${image.storageKey}`, alt: image.altText, width: image.width, height: image.height });

export async function serializeProduct(product) {
  const [garment, theme, rules] = await Promise.all([
    GarmentType.findOne({ slug: product.garmentSlug }).lean(),
    product.themeSlug ? Theme.findOne({ slug: product.themeSlug }).lean() : null,
    PricingRule.find({ garmentSlug: product.garmentSlug, productKind: product.category === "customized" ? "customized" : "standard" }).lean(),
  ]);
  const primary = product.images.find((item) => item.role === "primary");
  if (!primary) throw new Error(`Published product ${product.productId} has no primary image`);
  const hover = product.images.find((item) => item.role === "hover");
  const sizes = new Map(garment.sizes.map((item) => [item.slug, item]));
  const variants = [...product.variants].sort((a, b) => sizes.get(a.sizeSlug).displayOrder - sizes.get(b.sizeSlug).displayOrder);
  return {
    id: product.productId, slug: product.slug, name: product.name, category: product.category, garmentType: product.garmentSlug,
    theme: theme ? { id: theme.slug, label: theme.name } : null, description: product.description ?? null,
    images: { primary: imageResponse(primary), hover: hover ? imageResponse(hover) : null, gallery: product.images.filter((item) => item.role === "gallery").sort((a, b) => a.position - b.position).map(imageResponse) },
    colors: garment.colors.filter((item) => item.slug === product.colorSlug).map((item) => ({ id: item.slug, label: item.name, swatch: item.swatch ?? null })),
    sizes: variants.map((variant) => ({ id: variant.sizeSlug, label: sizes.get(variant.sizeSlug).name, swatch: null })),
    prices: priceResponse(rules),
    variants: variants.map((variant) => ({ id: variant.id, colorId: product.colorSlug, sizeId: variant.sizeSlug, availability: variant.availability, prices: variant.prices?.size ? Object.fromEntries([...variant.prices].map(([currency, amount]) => [currency, { minorAmount: amount, currency }])) : null })),
    availability: product.availability, displayOrder: product.displayOrder,
  };
}

export async function catalogOptions() {
  const [themes, garments, rules] = await Promise.all([Theme.find().sort({ displayOrder: 1 }).lean(), GarmentType.find().sort({ displayOrder: 1 }).lean(), PricingRule.find().lean()]);
  return { themes: themes.map((item) => ({ id: item.slug, label: item.name })), garments: garments.map((garment) => ({ id: garment.slug, label: garment.name, colors: garment.colors.map((item) => ({ id: item.slug, label: item.name, swatch: item.swatch ?? null })), sizes: garment.sizes.map((item) => ({ id: item.slug, label: item.name, swatch: null })), standardPrices: priceResponse(rules.filter((r) => r.garmentSlug === garment.slug && r.productKind === "standard")), customizedPrices: priceResponse(rules.filter((r) => r.garmentSlug === garment.slug && r.productKind === "customized")) })) };
}

export async function validateConfiguration(payload, currentId) {
  const [garment, theme, duplicate] = await Promise.all([
    GarmentType.findOne({ slug: payload.garment_slug }).lean(),
    payload.theme_slug ? Theme.findOne({ slug: payload.theme_slug }).lean() : null,
    Product.findOne({ slug: payload.slug, ...(currentId ? { productId: { $ne: currentId } } : {}) }).lean(),
  ]);
  if (!garment) throw new HttpError(422, "Unknown garment type");
  if (payload.category !== "customized" && payload.category !== garment.slug) throw new HttpError(422, "Standard product category must match its garment type");
  if (payload.category === "customized" && payload.theme_slug) throw new HttpError(422, "Customized products cannot use predefined themes");
  if (payload.category !== "customized" && !payload.theme_slug) throw new HttpError(422, "Standard products require a theme");
  if (payload.theme_slug && !theme) throw new HttpError(422, "Unknown theme");
  if (!garment.colors.some((item) => item.slug === payload.color_slug)) throw new HttpError(422, "Color is not available for this garment");
  const requested = [...new Set(payload.size_slugs)];
  if (requested.length !== payload.size_slugs.length || requested.some((slug) => !garment.sizes.some((item) => item.slug === slug))) throw new HttpError(422, "Size is not available for this garment");
  if (duplicate) throw new HttpError(422, "Product slug already exists");
  if (payload.images.filter((item) => item.role === "primary").length !== 1) throw new HttpError(422, "A product requires exactly one primary image");
  if (new Set(payload.images.map((item) => item.position)).size !== payload.images.length) throw new HttpError(422, "Image positions must be unique");
  if (payload.images.some((item) => !item.media_asset_id && !item.storage_key && !item.public_url)) throw new HttpError(422, "Every image requires a storage reference");
  return garment;
}

async function buildImages(images) {
  return Promise.all(images.map(async (image) => {
    const asset = image.media_asset_id ? await MediaAsset.findOne({ assetId: image.media_asset_id }).lean() : null;
    if (image.media_asset_id && !asset) throw new HttpError(422, "Unknown media asset");
    return { id: crypto.randomUUID(), mediaAssetId: asset?.assetId, storageKey: asset?.storageKey ?? image.storage_key, publicUrl: asset?.publicUrl ?? image.public_url, altText: image.alt_text, width: asset?.width ?? image.width, height: asset?.height ?? image.height, position: image.position, role: image.role };
  }));
}

export async function createProduct(payload) {
  const garment = await validateConfiguration(payload);
  const sizeOrder = new Map(garment.sizes.map((item) => [item.slug, item.displayOrder]));
  const product = await Product.create({ slug: payload.slug, name: payload.name, description: payload.description, category: payload.category, garmentSlug: payload.garment_slug, themeSlug: payload.theme_slug, colorSlug: payload.color_slug, availability: payload.availability, displayOrder: payload.display_order, images: await buildImages(payload.images), variants: [...payload.size_slugs].sort((a, b) => sizeOrder.get(a) - sizeOrder.get(b)).map((sizeSlug) => ({ id: crypto.randomUUID(), sizeSlug, availability: payload.availability })) });
  return serializeProduct(product);
}

export async function updateProduct(productId, payload) {
  const product = await Product.findOne({ productId });
  if (!product) throw new HttpError(404, "Product not found");
  const garment = await validateConfiguration(payload, productId);
  const existing = new Map(product.variants.map((item) => [item.sizeSlug, item.id]));
  const sizeOrder = new Map(garment.sizes.map((item) => [item.slug, item.displayOrder]));
  Object.assign(product, { slug: payload.slug, name: payload.name, description: payload.description, category: payload.category, garmentSlug: payload.garment_slug, themeSlug: payload.theme_slug, colorSlug: payload.color_slug, availability: payload.availability, displayOrder: payload.display_order, archivedAt: null, images: await buildImages(payload.images), variants: [...payload.size_slugs].sort((a, b) => sizeOrder.get(a) - sizeOrder.get(b)).map((sizeSlug) => ({ id: existing.get(sizeSlug) ?? crypto.randomUUID(), sizeSlug, availability: payload.availability })) });
  await product.save(); return serializeProduct(product);
}
